keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+     in-place Pod resources resizing.
+   * _Case 2_: ResizingPod.Status is false, and ResizingPod.Reason is
+     FailedNodeCapacity. The initiating actor may take alternative action.
+     For example, based on Retry policy, initiating actor (e.g VPA) may choose to:
+     - Evict the Pod to trigger a replacement Pod with updated resources,
+     - Do nothing, and let Kubelet backoff and retry in-place resize.
+
+#### Transitions of the ResizingPod PodCondition
+
+The following diagram shows possible transitions of ResizingPod.Status and
+ResizingPod.Reason fields respectively.
+
+```text
+
+   +-------------------------------------+
+   |                                     |
+   |                                    2|
+   |                                +----v----+
+   |                                |         |
+   |          +---------------------+  false  +-------------------------+
+   |          |                     | Success |                         |
+   |          |                     |         |                         |
+   |          |                     +----+----+                         |
+   |          |                          |                              |
+   |         1|                         3|                             5|
+   |   +------v-----+             +------v-----+             +----------v---------+
+   |   |            |             |            |             |                    |
+   |   |    true    |4            |    true    |7            |        false       |
+   +---+ InProgress <-------------+ PreEmpting <-------------+ FailedNodeCapacity |
+       |            |             |            |             |                    |
+       +------^-----+             +------------+             +----------+---------+
+             6|                                                         |
+              |                                                         |
+              +---------------------------------------------------------+
+
+```
+
+1. Kubelet, on initiating in-place resize.
+1. Kubelet, on successful completion of in-place resize.
+1. Kubelet, on initiating pre-emption of lower priority Pods.
+1. Kubelet, on completing pre-emption of lower priority Pods.
+1. Kubelet, on Node not having capacity to resize Pod.
+1. Kubelet, on initiating in-place resize retry.
+1. Kubelet, on initiating in-place resize retry with pre-emption.
+
+#### Container resource limit update ordering
+
+When in-place resize is desired for multiple Containers in a Pod, Kubelet updates
+resource limit for the Pod and its Containers in the following manner:
+  1. If resource resizing results in net-increase of a resource type (CPU or Memory),
+     Kubelet first updates Pod-level cgroup limit for the resource type, and then
+     updates the Container resource limit.
+  1. If resource resizing results in net-decrease of a resource type, Kubelet first
```

Comment:
as noted earlier, for cpu this is immediate.  for memory, the kubelet will need to induce pressure on the cgroup by setting a value based on its current usage.

Reply:
This would be handled by the runtime, no? What's the expected behavior for the runtime and what's the timeout duration for the CRI call?

Reply:
Yes this should be handled by run-time.

Main intent of this section is to clarify how the limit update (UpdateContainerResources CRI call) should be ordered when multiple containers are being resized in a request. For e.g if Pod sum(memory) = 5G with containers c1 (2G), c2 (3G), and a resize requests c1(4G), c2(2G), we should set pod limit to 6G first, then update c2 limit before c1.